test/CrossVersionTokenValidation.Tests/SamlCrossVersionTokenValidationTests.cs

```diff
@@ -0,0 +1,789 @@
+
+                    if (String.CompareOrdinal(authStatement1.IPAddress, authStatement2.IPAddress) != 0)
+                        localContext.Diffs.Add(Environment.NewLine + $"authStatement1.IPAddress != authStatement2.IPAddress: {authStatement1.IPAddress}, {authStatement2.IPAddress}");
+
+                    AreSamlSubjectsEqual(authStatement1.SamlSubject, authStatement2.Subject, localContext);
+                    AreSamlObjectEnumsEqual(authStatement1.AuthorityBindings, authStatement2.AuthorityBindings, localContext);
+                }
+                else if (statement1 is System.IdentityModel.Tokens.SamlAttributeStatement attributeStatement1)
+                {
+                    var attributeStatement2 = statement2 as SamlAttributeStatement;
+                    AreSamlSubjectsEqual(attributeStatement1.SamlSubject, attributeStatement2.Subject, localContext);
+                    AreSamlObjectEnumsEqual(attributeStatement1.Attributes, attributeStatement2.Attributes, localContext);
+                }
+                else if (statement1 is System.IdentityModel.Tokens.SamlAuthorizationDecisionStatement decisionStatement1)
+                {
+                    var decisionStatement2 = statement2 as SamlAuthorizationDecisionStatement;
+                    AreSamlSubjectsEqual(decisionStatement1.SamlSubject, decisionStatement2.Subject, localContext);
+                    AreSamlObjectEnumsEqual(decisionStatement1.SamlActions, decisionStatement2.Actions, localContext);
+
+                    if (String.CompareOrdinal(decisionStatement1.AccessDecision.ToString(), decisionStatement2.Decision) != 0)
+                        localContext.Diffs.Add(Environment.NewLine + $"decisionStatement1.AccessDecision != decisionStatement2.Decision: {decisionStatement1.AccessDecision}, {decisionStatement2.Decision}");
+
+                    AreSamlEvidencesEqual(decisionStatement1.Evidence, decisionStatement2.Evidence, localContext);
+
+                    if (String.CompareOrdinal(decisionStatement1.Resource, decisionStatement2.Resource) != 0)
+                        localContext.Diffs.Add(Environment.NewLine + $"decisionStatement1.Resource != decisionStatement2.Resource: {decisionStatement1.Resource}, {decisionStatement2.Resource}");
+                }
+            }
+
+            return context.Merge(localContext);
+        }
+
+        private static bool AreSamlObjectEnumsEqual<T1, T2>(ICollection<T1> objs1, ICollection<T2> objs2, CompareContext context)
+        {
+            if (objs1 == null && objs2 == null)
+                return true;
+
+            if (objs1 == null)
+            {
+                context.Diffs.Add(Environment.NewLine + "obj1 is null");
+                return false;
+            }
+
+            if (objs2 == null)
+            {
+                context.Diffs.Add(Environment.NewLine + "obj2 is null");
+                return false;
+            }
```

Comment:
You could break out these three checks into a separate function. then each method could use them.